src/storage/conversion/ChainedConverter.ts

```diff
@@ -1,89 +1,305 @@
+import { RepresentationConverter } from './RepresentationConverter';
 import type { RepresentationConverterArgs } from './RepresentationConverter';
-import { TypedRepresentationConverter } from './TypedRepresentationConverter';
+import type { TypedRepresentationConverter } from './TypedRepresentationConverter';
+
+type ConverterPreference = {
+  converter: TypedRepresentationConverter;
+  type: string;
+  weight: number;
+};
+
+type ConversionPath = {
+  converters: TypedRepresentationConverter[];
+  intermediateTypes: string[];
+  inTypes: ValuePreferences;
+  outTypes: ValuePreferences;
+};
+
+type MatchedPath = {
+  path: ConversionPath;
+  inType: string;
+  outType: string;
+  weight: number;
+};
+
+/**
+ * Finds the best media type that matches both the given `ValuePreferences`.
+ * It is assumed that `left` contains the more specific types than `right` if there are differences,
+ * since `left` usually corresponds to a converters output and `right` to the next converters input.
+ * "Best" implies the highest weight, which will also be returned.
+ */
+function getMatchingType(left: ValuePreferences, right: ValuePreferences): { type: string; weight: number } {
+  let bestMatch: { type: string; weight: number } = { type: 'invalid', weight: 0 };
+
+  // Try to find the matching type with the best weight
+  const rightEntries = Object.entries(right);
+  for (const [ leftType, leftWeight ] of Object.entries(left)) {
+    if (leftWeight <= bestMatch.weight) {
+      continue;
+    }
+    for (const [ rightType, rightWeight ] of rightEntries) {
+      const weight = leftWeight * rightWeight;
+      if (weight > bestMatch.weight && matchesMediaType(leftType, rightType)) {
+        bestMatch = { type: leftType, weight };
+        if (weight === 1) {
+          return bestMatch;
+        }
+      }
+    }
+  }
+
+  return bestMatch;
+}
+
+/**
+ * An LRU cache for storing `ConversionPath`s.
+ */
+class LruPathCache {
+  private readonly maxSize: number;
+  private readonly paths: ConversionPath[] = [];
+
+  public constructor(maxSize: number) {
+    this.maxSize = maxSize;
+  }
+
+  /**
+   * Add the given path to the cache as most recently used.
+   */
+  public add(path: ConversionPath): void {
+    this.paths.push(path);
+    if (this.paths.length > this.maxSize) {
+      this.paths.shift();
+    }
+  }
+
+  /**
+   * Find a path that can convert the given type to the given preferences.
+   * Note that this finds the first matching path in the cache,
+   * not the best one, should there be multiple results.
+   * In practice this should almost never be the case though.
+   */
+  public find(inType: string, outPreferences: ValuePreferences): MatchedPath | undefined {
+    // Last element has more chance of being the correct one
+    for (let i = this.paths.length - 1; i >= 0; --i) {
+      const path = this.paths[i];
+      const match = this.getMatchedPath(inType, outPreferences, path);
```

Comment:
I don't understand why this calculation has to happen if we're just reading from a cache. Add comment?